src/Mesh.cpp

```diff
@@ -182,21 +183,161 @@ double Mesh::gridUnitSI() const
 Mesh &Mesh::setGridUnitSI(double gusi)
 {
     setAttribute("gridUnitSI", gusi);
+    if (auto series_opt = retrieveSeries_optional(); series_opt.has_value())
+    {
+        if (auto version = series_opt->openPMD(); version >= "2.")
+        {
+            std::cerr << "[Mesh::setGridUnitSI] Warning: Setting a scalar "
+                         "`gridUnitSI` in a file with openPMD version '" +
+                    version +
+                    "'. Consider specifying a vector instead in order to "
+                    "specify "
+                    "the gridUnitSI per axis (ref.: "
+                    "https://github.com/openPMD/openPMD-standard/pull/193)."
+                      << std::endl;
+        }
+    }
+    return *this;
+}
+
+Mesh &Mesh::setGridUnitSI(std::vector<double> gusi)
+{
+    return setGridUnitSIPerDimension(std::move(gusi));
+}
+
+namespace
+{
+    uint64_t retrieveMeshDimensionality(Mesh const &m)
+    {
+        try
+        {
+            return m.axisLabels().size();
+        }
+        catch (no_such_attribute_error const &)
+        {
+            // no-op, continue with fallback below
+        }
+
+        // maybe we have record components and can ask them
+        if (auto it = m.begin(); it != m.end())
+        {
+            return it->second.getDimensionality();
+        }
+        /*
+         * Since some backends cannot distinguish between vector and
+         * scalar values, the most likely answer here is 1.
+         */
+        return 1;
+    }
+} // namespace
+
+std::vector<double> Mesh::gridUnitSIPerDimension() const
+{
+    return getAttribute("gridUnitSI").get<std::vector<double>>();
```

Comment:
TODO: introduce the same backwards-compatible logic as for `gridUnitSI` (line 330)